1-Draft/RFCNNNN-Side-By-Side-Support.md

```diff
@@ -0,0 +1,253 @@
+could specify a specific version that they want to install (e.g. `brew cask
+install powershell@6.2.0`) to install different versions of PowerShell
+side-by-side with one another, with the versions automatically being installed
+in an appropriate version folder (which they are already anyway). Chocolatey
+already has support for installing a specific version, but it does not seem to
+work at the moment (I tried installing PowerShell Core 6.2.0 and ended up with
+6.2.1, with an error indicating that 6.2.0 was not installed).
+
+The UI installers should provide options to upgrade the latest installed
+version of PowerShell or install a new version side-by-side.
+
+### Listing installed versions of PowerShell
+
+```bash
+# This launches the latest version of PowerShell, returns a list of installed
+# versions and their installation location, and then exits
+pwsh -ListVersions
+# This does the same thing
+pwsh -lv
+```
+
+```output
+6.2.1 [/usr/local/microsoft/powershell/6.2.1/]
+7.0.100-preview1 [/usr/local/microsoft/powershell/7.0.100-preview1]
+```
+
+```powershell
+
+# This returns a list of installed versions and their installation location as
+# objects
+Get-InstalledPSVersion
+```
+
+```output
+Version                    InstallLocation
+-------                    ---------------
+6.2.1                      C:\ProgramFiles\PowerShell\6.2.1
+7.0.100-preview1           C:\ProgramFiles\PowerShell\7.0.100-preview1
+```
+
+### Launching a specific version of PowerShell
+
+```bash
+# This launches the latest version of PowerShell, gets the installation
+# location of the specified version, launches pwsh for that version, and exits
+# when the launched version exits
+pwsh -Version 6.2.1
+# This does the same thing
+pwsh -v 6.2.1
+```
+
+```output
+# A new window with the desired version of PowerShell open.
+# If it is possible to do have this window on a new tab in the new Terminal,
+# that would be great.
+```
+
+## Specification
+
+* create a `Get-InstalledPSVersion [<CommonParameters>]` cmdlet that returns
+the installed versions of PowerShell along with their locations
+* update the `-version` argument of `pwsh.exe` to attempt to launch a specific
+version of PowerShell
+* add a `-ListVersions` argument to `pwsh.exe` that internally invokes
+`Get-InstalledPSVersion`, formats and outputs the results, and exits
+* update the release packaging process to produce versioned packages
+
+## Alternate Proposals and Considerations
+
+### Include Windows PowerShell version support
+
+It would be worth considering having `Get-InstalledPSVersion` return the
+installed version of Windows PowerShell along with its location, and having
+`pwsh -version` support launching Windows PowerShell. That would allow users to
+launch any installed PowerShell version from `pwsh.exe`, making it easy to set
+up automation with one command to invoke. Arguments from `pwsh.exe` could be
+mapped to arguments on `powershell.exe`.
```

Comment:
I think that `Get-InstalledPSVersion` should definitely include Windows PowerShell, but I don't think the rest is really necessary given the binaries are different names. But it's not an important point.